src/ArcgisAccount/accountManagerUtils.js

```diff
@@ -0,0 +1,239 @@
+      clientId: clientId,
+      portal: portal,
+      popup: popup
+    });
+
+    const account = await createAccountObject({ dSession, portal, clientId });
+
+    // Clear query string token from URL
+    window.history.replaceState({}, document.title, window.location.pathname);
+    return account;
+  } catch (err) {
+    console.error('Error getting User Session (completeOAuth).');
+    return null;
+  }
+};
+
+/** Complete auth and return account with serialized portal and session  */
+export const completeAuth = async ({ authProps }) => {
+  const { clientId, portalUrl, popup } = authProps || {};
+
+  const account = await completeOAuthSignIn({
+    clientId,
+    portalUrl,
+    popup
+  });
+
+  if (account) {
+    //Create key
+    const key = createAccountKey({ account });
+    //Set state
+
+    return { key: key, user: account };
+  } else {
+    return { key: undefined, user: undefined };
+  }
+};
+
+//** Get portal object */
+export const getPortal = async ({ portalUrl, session }) => {
+  try {
+    const portalObject = await getSelf({
+      portal: portalUrl,
+      authentication: session
+    });
+    return portalObject;
+  } catch (err) {
+    console.error('Error getting portal item.', err);
+    return null;
+  }
+};
+
+//** Login */
+export const loginOAuth2 = async ({
+  clientId,
+  redirectUri,
+  portalUrl,
+  popup,
+  name,
+  setAccountManagerState
+}) => {
+  beginOAuthSignIn({
+    clientId,
+    redirectUri,
+    portalUrl,
+    popup,
+    name,
+    setAccountManagerState
+  });
+};
+
+//** Logout */
+export const logoutOAuth2 = ({ url, clientId, token }) => {
+  //** Formal Logout requires esri_auth cookie */
+  //AGO
+  //return `${portalUrl.origin}/home/pages/Account/manage_accounts.html#redirect_uri=${encodedRedirectUrl}&client_id=${params.clientId}&signout=true`
+  //Enterprise
+  //return `${portalUrl}/oauth2/signout?client_id=${params.clientId}&redirect_uri=${encodedRedirectUrl}`;
+
+  //** Revoke Token (OAuth2): https://developers.arcgis.com/rest/users-groups-and-items/revoke-token.htm */
+  //Build logout request
+  if (!clientId || !token) {
+    return { success: false, error: 'Empty client_id / token' };
+  }
+  const revokeTokenEndpoint = `${url}/oauth2/revokeToken`;
+  const data = {
+    client_id: clientId,
+    auth_token: token
+  };
+
+  try {
+    request(revokeTokenEndpoint, {
+      params: data,
+      httpMethod: 'POST'
+    }).then(res => {
+      return { success: res.success };
+    });
+  } catch (err) {
+    return { success: false, error: err };
+  }
+};
+
+//** Get Org Thumbnail */
+export const getOrgThumbnail = props => {
+  const { session, portal, token } = props || {};
+  //org thumbnail (https://developers.arcgis.com/rest/users-groups-and-items/portal-self.htm)
+  let orgThumbnail = {
+    url: undefined,
+    letters: undefined
+  };
+  orgThumbnail.url =
+    portal && portal.thumbnail && session && session.portal && token
```

Comment:
Can we use optional chaining for this kind of stuff or does it break it? `portal?.thumbnail && session?.portal && token`

Reply:
I handled this by destructuring the nested props.